File: projects/06/test_parser.py
import unittest
import parser

class TestParser(unittest.TestCase):
    def setUp(self):
        self.test_parser = parser.Parser()

    def test_hasMoreCommands(self):
        self.assertFalse(self.test_parser.hasMoreCommands())

    def test_commandType(self):
        tests = ['@', '@123', '@TEST']
        for test in tests:
            self.assertEqual(self.test_parser.commandType(test), 'A_COMMAND', 'Failed: ' + test)

        tests = ['A=0' 'A=1', 'A=-1', 'A=D,' 'A=A', 'A=!D', 'A=!A', 'A=-D', 'A=-A', 'A=D+1', 'A=D-1', 'A=A-1', 'A=D+A', 'A=D-A', 
                 'A=A-D', 'A=D&A', 'A=D|A', 'A=M', 'A=!M', 'A=-M', 'A=M+1', 'A=D+M', 'A=D-M', 'A=M-D', 'A=D&M', 'A=D|M']
        for test in tests:
            self.assertEqual(self.test_parser.commandType(test), 'C_COMMAND', 'Failed: ' + test)

        tests = ['()', '(LOOP)', '(1)']
        for test in tests:
            self.assertEqual(self.test_parser.commandType(test), 'L_COMMAND', 'Failed: ' + test)
    
    def test_symbol(self):
        self.test_parser.command = '@123'
        self.assertEqual(self.test_parser.symbol(), '123')
        
        self.test_parser.command = '(LOOP)'
        self.assertEqual(self.test_parser.symbol(), 'LOOP')

    def test_dest(self):
        self.test_parser.command = 'A=0'
        self.assertEqual(self.test_parser.dest(), 'A')
        
        self.test_parser.command = 'D=A'
        self.assertEqual(self.test_parser.dest(), 'D')
    
    def test_comp(self):
        self.test_parser.command = 'A=M'
        self.assertEqual(self.test_parser.comp(), 'M')
        
        self.test_parser.command = 'A=A-1'
        self.assertEqual(self.test_parser.comp(), 'A-1')

    def test_jump(self):
        self.test_parser.command = 'D;JGT'
        self.assertEqual(self.test_parser.jump(), 'JGT')
        
        self.test_parser.command = 'D;JMP'
        self.assertEqual(self.test_parser.jump(), 'JMP')

if __name__ == "__main__":
    unittest.main()

File: projects/06/Assembler/src/TestParser.java
import junit.framework.*;
import org.junit.Test;
import java.util.*;

public class TestParser extends TestCase {
	
	/* Add.asm -> Add.hack
	 * 
	 * @2    -> 0000000000000010
     * D=A   -> 1110110000010000
     * @3    -> 0000000000000011
     * D=D+A -> 1110000010010000
     * @0    -> 0000000000000000
     * M=D   -> 1110001100001000
	 * 
	 * */

	public TestParser(String name) {
		super(name);
	}

	//Setup
	String testFile = "../add/Add.asm";
	Parser myParser = new Parser();	
	
	@Test
	public void testInput() {
		myParser.initialize(testFile);	
		ArrayList<String> testInput = new ArrayList<String>(Arrays.asList("@2", "D=A", "@3", "D=D+A", "@0", "M=D"));
		
		assertEquals(testInput, myParser.input);

	}
	
	@Test
	public void testHasMoreCommands() {
		myParser.initialize(testFile);
		assertTrue(myParser.hasMoreCommands());
	}
	
	@Test
	public void testAdvance() {
		myParser.initialize(testFile);
//		System.out.print(myParser.input);
		
		myParser.advance();
		assertEquals("@2", myParser.currentCommand);
		
		myParser.advance();
		assertEquals("D=A", myParser.currentCommand);

		
		myParser.advance();
		assertEquals("@3", myParser.currentCommand);

		
		myParser.advance();
		assertEquals("D=D+A", myParser.currentCommand);

		
		myParser.advance();
		assertEquals("@0", myParser.currentCommand);

		
		myParser.advance();
		assertEquals("M=D", myParser.currentCommand);

	}
	
	@Test
	public void testCommandType() {
		myParser.initialize(testFile);
		
		myParser.currentCommand = "@2";
		assertEquals("A_COMMAND", myParser.commandType());
		
		myParser.currentCommand = "@0";
		assertEquals("A_COMMAND", myParser.commandType());
		
		myParser.currentCommand = "D=A";
		assertEquals("C_COMMAND", myParser.commandType());
		
		myParser.currentCommand = "D=D+A";
		assertEquals("C_COMMAND", myParser.commandType());
		
		myParser.currentCommand = "(LOOP)";
		assertEquals("L_COMMAND", myParser.commandType());		
	}
	
	@Test
	public void testSymbol() {
		myParser.initialize(testFile);
		
		myParser.currentCommand = "@2";
		assertEquals("2", myParser.symbol());
		
		myParser.currentCommand = "(LOOP)";
		assertEquals("LOOP", myParser.symbol());
	}
	
	@Test
	public void testDest() {
		myParser.initialize(testFile);
		
		myParser.currentCommand = "D=A";
		assertEquals("D", myParser.dest());
		
		myParser.currentCommand = "D=D+A";
		assertEquals("D", myParser.dest());
		
		myParser.currentCommand = "M=D";
		assertEquals("M", myParser.dest());
	}
	
	@Test
	public void testComp() {
		myParser.initialize(testFile);
		
		myParser.currentCommand = "D=A";
		assertEquals("A", myParser.comp());
		
		myParser.currentCommand = "D=D+A";
		assertEquals("D+A", myParser.comp());
		
		myParser.currentCommand = "M=D";
		assertEquals("D", myParser.comp());

	}
	
	@Test
	public void testJump() {
		myParser.initialize(testFile);
		
		myParser.currentCommand = "0;JMP";
		assertEquals("JMP", myParser.jump());
		
		myParser.currentCommand = " D;JLE";
		assertEquals("JLE", myParser.jump());
		
	}
	
//	@Test
//	public void testDest() {
//		myParser.initialize(testFile);
//		myParser.currentCommand = "null";
//		assertEquals("000", myParser.dest());
//	}

}